File: files/CH-140.py
import pandas as pd
from pandas.tseries.offsets import Day

path = "CH-140 Golden Period.xlsx"
input = pd.read_excel(path, usecols="B:D", skiprows=1, nrows=25)
test = pd.read_excel(path, usecols="F:H", skiprows=1, nrows=3).rename(columns=lambda x: x.split('.')[0])

result = input.groupby(['Date', 'Customer'], as_index=False)['Qty'].sum()
result = (result.set_index('Date')
          .groupby('Customer')
          .apply(lambda x: x.reindex(pd.date_range(x.index.min(), "2024-11-01", freq='D')))
          .reset_index(level=0, drop=True)
          .reset_index())
result['Qty'] = result['Qty'].fillna(0)
result['Customer'] = result['Customer'].ffill()
result['rolling_sum'] = result.groupby('Customer')['Qty'].rolling(window=10, min_periods=10).sum().reset_index(level=0, drop=True)
result = result[result.groupby('Customer')['rolling_sum'].transform('max') == result['rolling_sum']]
result = result.groupby('Customer').tail(1)
result['min_date'] = (result['index'] - Day(9)).dt.strftime('%y-%m-%d')
result['Date'] = result['index'].dt.strftime('%y-%m-%d')
result['Period'] = result['min_date'] + ' to ' + result['Date']
result = result[['Customer', 'Period', 'rolling_sum']].rename(columns={'rolling_sum': 'Total Qty'}).reset_index(drop=True)
result['Total Qty'] = result['Total Qty'].astype('int64')

print(result.equals(test)) # True

File: files/100-199/140/CH-140.R
library(tidyverse)
library(readxl)
library(slider)

path = "files/CH-140 Golden Period.xlsx"
input = read_excel(path, range = "B2:D26")
test  = read_excel(path, range = "F2:H5")

result = input %>%
  summarise(Qty = sum(Qty), .by = c(Date, Customer)) %>%
  group_by(Customer) %>%
  complete(Date = seq.Date(min(as.Date(Date)), as.Date("2024/11/01"), by = "1 day")) %>%
  left_join(input)  %>%
  replace_na(list(Qty = 0)) %>%
  group_by(Customer) %>%
  mutate(rolling_sum = slide_dbl(Qty, sum, .before = 9, .complete = TRUE)) %>%
  filter(rolling_sum == max(rolling_sum, na.rm = T)) %>%
  filter(Date == max(Date, na.rm = T)) %>% 
  mutate(min_date = Date - days(9)) %>%
  select(Customer, min_date, Date, `Total Qty` = rolling_sum) %>%
  mutate(min_date = format(min_date, "%y-%m-%d"), Date = format(Date, "%y-%m-%d")) %>%
  unite("Period", min_date, Date, sep = " to ")

all.equal(result, test, check.attributes = F)
#> [1] TRUE
